test/aleph/http/multipart_test.clj

```diff
@@ -188,52 +189,132 @@
 
     (.close ^java.io.Closeable s)))
 
-(defn- pack-chunk [{:keys [content] :as chunk}]
-  (cond-> (dissoc chunk :file)
-    (not (string? content))
-    (dissoc :content)))
-
-(defn- decode-handler [req]
-  (let [chunks (-> req
-                   mp/decode-request
-                   s/stream->seq)]
-    {:status 200
-     :body (pr-str (map pack-chunk chunks))}))
-
-(defn- test-decoder [port url raw-stream?]
-  (let [s (http/start-server decode-handler {:port port
-                                             :shutdown-timeout 0
-                                             :raw-stream? raw-stream?})
-        chunks (-> (http/post url {:multipart parts})
-                   (deref 1e3 {:body "timeout"})
-                   :body
-                   bs/to-string
-                   clojure.edn/read-string
-                   vec)]
-    (is (= 6 (count chunks)))
-
-    ;; part-names
-    (is (= (map :part-name parts)
-           (map :part-name chunks)))
-
-    ;; content
-    (is (= "CONTENT1" (get-in chunks [0 :content])))
-
-    ;; mime type
-    (is (= "text/plain" (get-in chunks [2 :mime-type])))
-    (is (= "application/png" (get-in chunks [3 :mime-type])))
-
-    ;; filename
-    (is (= "file.txt" (get-in chunks [3 :name])))
-    (is (= "file.txt" (get-in chunks [4 :name])))
-
-    ;; charset
-    (is (= "ISO-8859-1" (get-in chunks [5 :charset])))
+(defn- pack-part [{:keys [content file] :as part}]
+  (cond-> part
+    (some? file)
+    (assoc :file (.getAbsolutePath ^java.io.File file))
+    (some? content)
+    (update :content bs/to-string)))
 
-    (.close ^java.io.Closeable s)))
+(defn- make-decode-handler [options]
+  (fn [req]
+    (let [parts (mp/decode-request req options)
+          body  (mapv (fn [{:keys [file] :as part}]
+                        (Thread/sleep 20) ;; wait a bit to ensure the files are cleaned when using automatic resource cleanup
+                        (let [packed-part (pack-part part)]
+                          (when (some? file)
+                            (is (instance? java.io.File file))
+                            (is (not (.exists ^java.io.File file)))) ;; demonstrates there is an issue when the stream is closed before accessing the file.
+                          packed-part))
+                      (s/stream->seq parts))]
+      {:status 200
+       :body (pr-str body)})))
+
+(defn- make-decode-manual-handler [options]
+  (fn [req]
+    (let [[parts cleanup-fn] (mp/decode-request req options)
+          body  (mapv (fn [{:keys [file] :as part}]
+                        (Thread/sleep 20) ;; wait a bit to ensure the files are cleaned when using automatic resource cleanup
+                        (let [packed-part (pack-part part)]
+                          (when (some? file)
```

Comment:
Trivial: `when-some`

Reply:
I will use `when file` instead, `when-some` imposes you to create a binding

Reply:
Ahh, right, right. A pattern-match failure of the brain!